agent/base

```diff
@@ -209,6 +209,9 @@ function verify_tool_group {
         printf -- "\t${script_name}: invalid ${param_name} option (\"${group}\"), directory not found: ${tools_group_dir}\n" >&2
         return 1
     fi
+    if [[ -z "$(ls -A ${tools_group_dir})" ]]; then
+        printf -- "\tWARNING: No tools are registered\n" >&2
```

Comment:
The similar `printf` at 209 uses `${script_name}:` and doesn't have an explicit severity. Including the script name seems to be a common convention that might be worth following.

Additionally, is a simple warning `printf` that continues operation sufficient here? It may give an indirect clue of why something bad happens later, but won't serve to prevent it.

Reply:
> The similar `printf` at 209 uses `${script_name}:` and doesn't have an explicit severity. Including the script name seems to be a common convention that might be worth following.

Yes I should have added `${script_name}:` here as well.

> Additionally, is a simple warning `printf` that continues operation sufficient here? It may give an indirect clue of why something bad happens later, but won't serve to prevent it.

We should't exit if no tools are registered, it is a valid use case to run user benchmark with no tools registered. However, I felt it ought to give some indication or warning when starting the benchmark that there are currently no tools registered on this host (and if something bad happens this may be the reason although its not in the message and I dont know how I should add that message). Also if you look at the comment from peter on the issue https://github.com/distributed-system-analysis/pbench/issues/2433 he seem to suggest that we should warn the user when no tools are found. 

Just a warning is definitely not sufficient and we need to handle exceptions more robustly when no tools are registered but that's another issue.

Reply:
The `-A` switch is an interesting choice:  if someone puts a "dot file" in the tools directory, that will silence this warning without "being a tool", right?  Is that a bug or a feature?

Not knowing the history or culture behind this setup, I would be inclined to omit the `-A` switch an let "dot files" stay hidden and ignored.

As for the greater question, I think we should allow users to run workloads without data collection tools (i.e., the workload will probably report its own results), but we _should_ warn them, because running a benchmark might be an expensive operation and we don't want them to unknowingly fail to collect all of their data.

Reply:
> Just a warning is definitely not sufficient and we need to handle exceptions more robustly when no tools are registered but that's another issue.

Fine; I was reading this in the context of your recent #2603 ... however you're both right we don't necessarily want to prevent using a benchmark script without explicit tools and a warning is likely appropriate. I'd like to see the message made more consistent, and I suspect Webb's comment about dropping the `-A` is a good one.